### ISSUE TITLE
[go_router] go_router 9.0.0 migration does not work for global listening location changes

### ISSUE THREAD

**Issue by cenumi** [flutter/flutter#129944]
### Is there an existing issue for this?

- [X] I have searched the [existing issues](https://github.com/flutter/flutter/issues)
- [X] I have read the [guide to filing a bug](https://flutter.dev/docs/resources/bug-reports)

### Steps to reproduce

go_router v9.0.0 removes `GoRouter.location`, which has extra logic is different from 'GoRouterState.location' and 'GoRouterDelegate.currentConfiguration.uri.toString()', the migration is not valid.

The [commit](https://github.com/flutter/packages/commit/6b70804799a95b18bed725cad6b2b651514c40a1) removes `_handleStateMayChange()` 

Steps:

1. run given example
2. click push /a button

### Expected results

GoRouterState /
RouterDelegate /
GoRouter /
GoRouterState /a
RouterDelegate /a
GoRouter /a

### Actual results

GoRouterState /
RouterDelegate /
GoRouter /
GoRouterState /
RouterDelegate /
GoRouter /a

### Code sample

<details><summary>Code sample</summary>

```dart
import 'dart:async';

import 'package:flutter/material.dart';
import 'package:go_router/go_router.dart';

void main() {
  runApp(const MyApp());
}

final router = GoRouter(
  routes: [
    GoRoute(
      path: '/',
      builder: (_, __) => const MyWidget(titleText: '/'),
      routes: [
        GoRoute(path: 'a', builder: (_, __) => const MyWidget(titleText: '/a')),
        GoRoute(path: 'b', builder: (_, __) => const MyWidget(titleText: '/b')),
      ],
    ),
  ],
);

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp.router(
      routerConfig: router,
    );
  }
}

class MyWidget extends StatefulWidget {
  const MyWidget({super.key, required this.titleText});
  final String titleText;

  @override
  State<MyWidget> createState() => _MyWidgetState();
}

class _MyWidgetState extends State<MyWidget> {
  Timer? _timer;

  @override
  void initState() {
    super.initState();
    if (widget.titleText == '/') {
      _timer = Timer.periodic(const Duration(seconds: 3), (timer) {
        _listenRouter();
      });
    }
  }

  void _listenRouter() {
    print('GoRouterState ${GoRouterState.of(context).location}');
    print('RouterDelegate ${GoRouter.of(context).routerDelegate.currentConfiguration.uri.toString()}');
    print('GoRouter ${GoRouter.of(context).location}');
  }

  @override
  void dispose() {
    _timer?.cancel();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: Text(widget.titleText)),
      body: SizedBox.expand(
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.center,
          children: [
            Text(
              'Current Location: ${GoRouterState.of(context).location}',
              style: Theme.of(context).textTheme.headlineMedium,
            ),
            FilledButton(
              onPressed: () {
                context.push('/a');
              },
              child: const Text('Push /a'),
            ),
            FilledButton(
              onPressed: () {
                context.push('/b');
              },
              child: const Text('Push /b'),
            ),
          ],
        ),
      ),
    );
  }
}

```

</details>

### Screenshots or Video

<details>
<summary>Screenshots / Video demonstration</summary>

[Upload media here]

</details>

### Logs

<details><summary>Logs</summary>

```console
[Paste your logs here]
```

</details>

### Flutter Doctor output

<details><summary>Doctor output</summary>

```console
Dartpad.dev stable Flutter 3.10.5 Dart SDK 3.0.5
go_router 8.0.3
```

</details>

**Comment by danagbemava-nc** [comment]
Hi @cenumi, below is what I see when I run the code sample you provided with go_router 8.0.3, since go_router 9 does not have location, I only see `/`. 

Is there something I'm missing?

<details>
<summary>logs</summary>


### go_router: 8.0.3

```
Launching lib/main.dart on macOS in debug mode...
2023-07-04 09:33:28.537 xcodebuild[2570:1122981] DVTCoreDeviceEnabledState: DVTCoreDeviceEnabledState_Disabled set via user default (DVTEnableCoreDevice=disabled)
--- xcodebuild: WARNING: Using the first of multiple matching destinations:
{ platform:macOS, arch:arm64, id:00008103-000329201AD1001E }
{ platform:macOS, arch:x86_64, id:00008103-000329201AD1001E }
warning: Run script build phase 'Run Script' will be run during every build because it does not specify any outputs. To address this warning, either add output dependencies to the script phase, or configure it to run in every build by unchecking "Based on dependency analysis" in the script phase. (in target 'Flutter Assemble' from project 'Runner')
Connecting to VM Service at ws://127.0.0.1:52962/0G6_HcYqMPM=/ws
flutter: GoRouterState /
flutter: RouterDelegate /
flutter: GoRouter /
flutter: GoRouterState /
flutter: RouterDelegate /
flutter: GoRouter /a
flutter: GoRouterState /
flutter: RouterDelegate /
flutter: GoRouter /a
flutter: GoRouterState /
flutter: RouterDelegate /
flutter: GoRouter /a
Application finished.
Exited
```
</details>

**Comment by cenumi** [comment]
> Hi @cenumi, below is what I see when I run the code sample you provided with go_router 8.0.3, since go_router 9 does not have location, I only see `/`.
> 
> Is there something I'm missing?


That's correct, so we can't get current location "/a" in '_timer', refer to [migrating doc](https://flutter.dev/go/go-router-v9-breaking-changes), it should be equivalent.

**Comment by danagbemava-nc** [comment]
Hi @cenumi, thanks for the clarification. 

I can reproduce the bug using the code sample provided above. 

With the current version of go_router, the correct path is not printed in the console. Only `/` is printed, even after navigating several times. 

<details>
<summary>flutter doctor -v</summary>

```
[!] Flutter (Channel stable, 3.10.5, on macOS 13.4.1 22F82 darwin-arm64, locale en-GB)
    • Flutter version 3.10.5 on channel stable at /Users/nexus/dev/sdks/flutter
    ! Warning: `flutter` on your path resolves to /Users/nexus/dev/sdks/flutters/bin/flutter, which is not inside your current Flutter SDK checkout at /Users/nexus/dev/sdks/flutter. Consider adding /Users/nexus/dev/sdks/flutter/bin to the front of your path.
    ! Warning: `dart` on your path resolves to /Users/nexus/dev/sdks/flutters/bin/dart, which is not inside your current Flutter SDK checkout at /Users/nexus/dev/sdks/flutter. Consider adding /Users/nexus/dev/sdks/flutter/bin to the front of your path.
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 796c8ef792 (3 weeks ago), 2023-06-13 15:51:02 -0700
    • Engine revision 45f6e00911
    • Dart version 3.0.5
    • DevTools version 2.23.1
    • If those were intentional, you can disregard the above warnings; however it is recommended to use "git" directly to perform update checks and upgrades.

[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.0)
    • Android SDK at /Users/nexus/Library/Android/sdk
    • Platform android-33, build-tools 33.0.0
    • Java binary at: /Users/nexus/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/222.4459.24.2221.9971841/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b802.4-9586694)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.3)
    • Xcode at /Applications/Xcode-14.3.0.app/Contents/Developer
    • Build 14E222b
    • CocoaPods version 1.12.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2022.2)
    • Android Studio at /Users/nexus/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/222.4459.24.2221.9971841/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b802.4-9586694)

[✓] Android Studio (version 2022.2)
    • Android Studio at /Users/nexus/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/222.4459.24.2221.10121639/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b802.4-9586694)

[✓] IntelliJ IDEA Ultimate Edition (version 2023.1.3)
    • IntelliJ at /Users/nexus/Applications/JetBrains Toolbox/IntelliJ IDEA Ultimate.app
    • Flutter plugin version 74.0.4
    • Dart plugin version 231.9161.14

[✓] IntelliJ IDEA Ultimate Edition (version 2023.1.2)
    • IntelliJ at /Users/nexus/Library/Application Support/JetBrains/Toolbox/apps/IDEA-U/ch-0/231.9011.34/IntelliJ IDEA.app
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart

[✓] IntelliJ IDEA Ultimate Edition (version 2023.1.3)
    • IntelliJ at /Users/nexus/Library/Application Support/JetBrains/Toolbox/apps/IDEA-U/ch-0/231.9161.38/IntelliJ IDEA.app
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart

[✓] VS Code (version 1.79.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.68.0

[✓] Connected device (2 available)
    • macOS (desktop) • macos  • darwin-arm64   • macOS 13.4.1 22F82 darwin-arm64
    • Chrome (web)    • chrome • web-javascript • Google Chrome 114.0.5735.198

[✓] Network resources
    • All expected network resources are available.

! Doctor found issues in 1 category.
```

```
[✓] Flutter (Channel master, 3.12.0-13.0.pre.42, on macOS 13.4.1 22F82 darwin-arm64, locale en-GB)
    • Flutter version 3.12.0-13.0.pre.42 on channel master at /Users/nexus/dev/sdks/flutters
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 5b12b7467f (13 hours ago), 2023-07-04 10:12:49 -0700
    • Engine revision 987b621eac
    • Dart version 3.1.0 (build 3.1.0-262.0.dev)
    • DevTools version 2.25.0

[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.0)
    • Android SDK at /Users/nexus/Library/Android/sdk
    • Platform android-33, build-tools 33.0.0
    • Java binary at: /Users/nexus/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/222.4459.24.2221.9971841/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b802.4-9586694)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.3)
    • Xcode at /Applications/Xcode-14.3.0.app/Contents/Developer
    • Build 14E222b
    • CocoaPods version 1.12.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2022.2)
    • Android Studio at /Users/nexus/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/222.4459.24.2221.9971841/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b802.4-9586694)

[✓] Android Studio (version 2022.2)
    • Android Studio at /Users/nexus/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/222.4459.24.2221.10121639/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b802.4-9586694)

[✓] IntelliJ IDEA Ultimate Edition (version 2023.1.3)
    • IntelliJ at /Users/nexus/Applications/JetBrains Toolbox/IntelliJ IDEA Ultimate.app
    • Flutter plugin version 74.0.4
    • Dart plugin version 231.9161.14

[✓] IntelliJ IDEA Ultimate Edition (version 2023.1.2)
    • IntelliJ at /Users/nexus/Library/Application Support/JetBrains/Toolbox/apps/IDEA-U/ch-0/231.9011.34/IntelliJ IDEA.app
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart

[✓] IntelliJ IDEA Ultimate Edition (version 2023.1.3)
    • IntelliJ at /Users/nexus/Library/Application Support/JetBrains/Toolbox/apps/IDEA-U/ch-0/231.9161.38/IntelliJ IDEA.app
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart

[✓] VS Code (version 1.79.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.68.0

[✓] Connected device (2 available)
    • macOS (desktop) • macos  • darwin-arm64   • macOS 13.4.1 22F82 darwin-arm64
    • Chrome (web)    • chrome • web-javascript • Google Chrome 114.0.5735.198

[✓] Network resources
    • All expected network resources are available.

• No issues found!
```

</details>

**Comment by a1573595** [comment]
I need GoRouter location listener to complete my crashlytics and analyze reports, but GoRouter.routerDelegate.addListener does not work like GoRouter.addListener.

**Comment by gengiscb** [comment]
this old code with 8.0
`void goRouterPopUntil(BuildContext context, String location) {
  while (GoRouter.of(context).location != location) {
    if (!context.canPop()) {
      return;
    }
    context.pop();
  }
}`

Migrated to 9.0.0 according to docs now does not works 
`void goRouterPopUntil(BuildContext context, String location) {
  while (GoRouterState.of(context).location != location) {
    if (!Navigator.of(context).canPop()) {
      return;
    }
    context.pop();
  }
}`

**Comment by buchen** [comment]
Adding my use case to @a1573595 [comment above](https://github.com/flutter/flutter/issues/129944#issuecomment-1621344981):
* I am using a StatefulShellRoute with a couple StatefulShellBranches; the listener is enabling landscape or portrait mode depending on the current location (some view support landscape view, most of them do not)
* ```GoRouter.routerDelegate.addListener ``` is not exhibiting the same behavior: apparently the listener is not fired for sub-routes in the StatefulShellBranch, e.g. there is an event for /x but not for /x/a

**Comment by chunhtai** [comment]
This is due to the currentConfiguration.uri does not report imperative route. I will update the migration guide to be more clearer on how to handle imperative route uri

**Comment by chunhtai** [comment]
Closing this issue.

I updated migration doc [flutter.dev/go/go-router-v9-breaking-changes](http://flutter.dev/go/go-router-v9-breaking-changes). The workaround is not too pretty, but it should work.

The reason I would like to move away from exposing such state in GoRouter is to avoid people doing GoRouter.of(context).location and expect it to rebuild when location change.

At some point, I would like to create a better system to listen to navigation event above GoRouter.
